blog/_posts/2019-03-19-debuggers.md

```diff
@@ -0,0 +1,254 @@
+---
+layout: post
+title: "A Julia interpreter and debugger"
+author: <a href="https://github.com/timholy">Tim Holy</a>
+author: <a href="https://github.com/KristofferC">Kristoffer Carlsson</a>
+author: <a href="https://github.com/pfitzseb">Sebastian Pfiztner</a>
+author: <a href="https://github.com/Keno">Keno Fischer</a>
+---
+
+The authors are pleased to announce the immediate availability of a
+debugger for the Julia programming language. The core infrastructure
+is in [JuliaInterpreter][JI].  Currently there are also three separate
+front-ends: a full-featured IDE in [Juno][], a REPL (console) GUI-like
+mode in [Rebugger][RB], and "step/next/continue" console interface in
+[Debugger][DB]. The new debugging infrastructure allows you to move
+stepwise through code, but you can also set breakpoints and trap
+errors. Moreover, the new debugging capabilities have been written to
+integrate seamlessly with [Revise][RV], so that you can continuously
+analyze and modify code in a single session.  Indeed, Revise has been
+rewritten with JuliaInterpreter at its foundation, and this has given
+both Revise and Rebugger powerful new capabilities. To better support
+a multitude of consumers, Revise has been split into two packages, and
+now offers a convenient "query" API in the [CodeTracking][CT] package
+to allow other packages to ask questions about the code in your
+running Julia session.
```

Comment:
Can we make this first paragraph a bit more exciting and results-oriented for a potential debugger-user?  Here's a go:

> We are thrilled to annouce the release of a fully-featured debugger for Julia.  You can now easily debug and introspect any Julia code through a variety of interfaces:
> * Step into functions and manually walk through your code while introspecting its state
> * Set breakpoints and trap errors, allowing you to discover what went wrong at the point of trouble
> * Interactively update and replace existing code to rapidly fix bugs in place without restarting
> * Or use the full-featured IDE in [Juno][] to bundle all these features together in an easy to use interface
>
> The debugger is itself a collection of tools that enable all those features.  The core is powered by [an interpreter][JI] that can faithfully run all Julia code and allows various front-ends to control its execution.  Each front-end is its own package: [Juno][] incorporates the debugger into its IDE, the [Rebugger][RB] provides a REPL text UI, and the traditional step/next/continue command-line interface is provided by [Debugger][DB].  All these new debugging capabilities seamlessly integrate with [Revise][RV], so that you can continuously
analyze and modify code in a single session.  

I very much like the last two sentences ("Indeed, Revise..." to the end), but I think they could be moved down to Revise's section.  The capability is awesome but the code structure and API is not what (I think) the average reader will care very much about.

Reply:
I'm stealing that :smile: